### DISCUSSION TITLE
Migration error: Unable to create a 'DbContext' of type 'AppDbContext'. The exception 'The entity type 'User' requires a primary key to be defined. If you intend ed to use a keyless entity type, call 'HasNoKey' in 'OnModelCreating'. For more information on keyless entity types, see https://go.microsoft.co m/fwlink/?linkid=2141943.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728

### DISCUSSION THREAD

**Discussion by JorgeSidney** [balta-io/forum#723] [Tira Dúvidas] [answered]
Ao entrar na aula de Identity Migrations, depois de rodar os comandos dotnet clean, dotnet build (sem erros) e dotnet ef migrations add v2, surgiu esse erro:
Unable to create a 'DbContext' of type 'AppDbContext'. The exception 'The entity type 'User' requires a primary key to be defined. If you intend
ed to use a keyless entity type, call 'HasNoKey' in 'OnModelCreating'. For more information on keyless entity types, see https://go.microsoft.co
m/fwlink/?linkid=2141943.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728

**Comment by henriqueweege** [comment]
Fala, @JorgeSidney ! Tudo bem?
Viu, teria como verificar se a classe User está herdando de IdentityUser<long>?

```csharp
namespace Dima.Api.Models;

public class User : IdentityUser<long>
{
    public List<IdentityRole<long>>? Roles { get; set; }
}
```

Caso sim, teria como verificar se o mapeamento está sendo feito certo na classe IdentityUserMapping? Dentro do método Configure deveria ser chamado o método .HasKey(u => u.Id).

```csharp
public class IdentityUserMapping : IEntityTypeConfiguration<User>
{
    public void Configure(EntityTypeBuilder<User> builder)
    {
        ...
        builder.HasKey(u => u.Id);
        ...
    }
}
```


Caso esteja tudo em ordem, teria como subir teu código para o GitHub para eu dar uma olhada?


Valeu!!

**Reply by JorgeSidney** [reply]
Olá Henrique. Poxa, muito obrigado aí pela dedicação meu amigo.
Pensei que só teria resposta ano que vem kkk.
Verifiquei ambos e parecem ok. 
Compartilhei o código contigo. Certamente foi coisa do newbie aqui.
Abraços.
Jorge.

**Reply by henriqueweege** [reply]
Fala, @JorgeSidney !
 
Capaz, sempre um prazer ajudar :) 

Olhei aqui, e o problema na verdade está no relacionamento entre a entidade Client e a entidade User do projeto Core.

No curso FullStack .NET do Balta a classe User do Core é apenas utilizada para mapeamentos no projeto Web, e por não se tratar de uma representação de uma tabela no banco, não há como utilizá-la para navegações do EF.

Para que isso seja possível, é necessário adicionar Id a essa classe (dai que vem a mensagem de erro ao gerar migration...) assim como as tratativas necessárias de mapeamento.

Não sei o quão boa prática é ter duas classes sendo mapeada para a mesma tabela no banco... avaliar prós e contras é uma boa ^-^

Valeu e feliz 2025!!

**Reply by JorgeSidney** [reply, marked as answer]
Olá Henrique. Feliz 2025 para você também!

Muito obrigado. 
Como disse, coisa de newbie aqui.
Eu simplesmente queria aproveitar o curso para ver se entendia o relacionamento entre tabelas, e queria ter colocado como User pode ter muitos Clients e Clients pode ter muitos Measurements. Na verdade, se eu pudesse concluir este relacionamento e manter User como no curso, seria o mundo ideal.
Um colega sugeriu a alteração e agora eu realmente estou tentando fazer funcionar como estava, pelo menos para gerar nova migration e continuar curso.

Mais uma vez, muito obrigado.
